File: typescript/src/set.spec.ts
describe("Set", () => {
    describe('空のSetを用意する', () => {
        let sut: Set<number> = new Set();

        it('Setは空である', () => {
            expect(sut.size).toEqual(0);
        });

        describe('値を追加する', () => {
            beforeAll(() => {
                sut.add(1);
                sut.add(2);
                sut.add(3);
            });

            it('要素数=3', () => {
                expect(sut.size).toEqual(3);
            });

            it('追加した値が格納されている', () => {
                expect(sut.has(1)).toBeTruthy();
                expect(sut.has(2)).toBeTruthy();
                expect(sut.has(3)).toBeTruthy();
            });

            describe('要素を合計する', () => {
                let sum = 0;

                beforeAll(() => {
                    for (let value of sut) {
                        sum += value;
                    }
                });

                it('結果は6', () => {
                    expect(sum).toEqual(6);
                });
            })
        });
    });

    describe('最初から値ありのSetを生成', () => {
        let sut = new Set([4, 5, 6]);

        it('要素数=3', () => {
            expect(sut.size).toEqual(3);
        });

        it('指定した値が格納されている', () => {
            expect(sut.has(4)).toBeTruthy();
            expect(sut.has(5)).toBeTruthy();
            expect(sut.has(6)).toBeTruthy();
        });
    });

    describe('集合演算', () => {
        const a = new Set([1, 2, 3, 4]);
        const b = new Set([3, 4, 5, 6]);

        describe('和集合', () => {
            let result: Set<number> = new Set();
            const expected = new Set([1, 2, 3, 4, 5, 6]);

            beforeAll(() => {
                for (let value of a) {
                    result.add(value);
                }

                for (let value of b) {
                    result.add(value);
                }
            });

            it('想定どおりの結果を得られる', () => {
                expect(result).toEqual(expected);
            });
        });

        describe('積集合', () => {
            let result: Set<number> = new Set();
            const expected = new Set([3, 4]);

            beforeAll(() => {
                result = new Set([...a].filter(e => (b.has(e))));
            });

            it('想定どおりの結果を得られる', () => {
                expect(result).toEqual(expected);
            });
        });

        describe('差集合', () => {
            let result: Set<number> = new Set();
            const expected = new Set([1, 2]);

            beforeAll(() => {
                result = new Set([...a].filter(e => (!b.has(e))));
            });

            it('想定どおりの結果を得られる', () => {
                expect(result).toEqual(expected);
            });
        });
    });
});

File: ruby/spec/set_spec.rb
# frozen_string_literal: true

require 'set'

describe Set do
  context '空のSetを用意する' do
    Given(:sut) { Set.new }
    Then { sut.empty? }

    context '新規の値を追加する' do
      When { sut << 1 }
      Then { sut.size == 1 }

      context '重複する値を追加する' do
        When { sut << 1 }
        Then { sut.size == 1 }
      end

      context '他の値も追加' do
        When { sut << 2 << 3 }
        Then { sut.size == 3 }
        Then { sut.include?(1) }
        Then { sut.include?(2) }
        Then { sut.include?(3) }

        context '要素を合計する' do
          When(:sum) do
            result = 0
            sut.each do |value|
              result += value
            end
            result
          end
          Then { sum == 6 }
        end
      end
    end
  end

  context '最初から値ありのSetを生成' do
    Given(:sut) { Set.new([4, 5, 6]) }
    Then { sut.size == 3 }
  end

  context '集合演算' do
    Given(:a) { Set.new([1, 2, 3, 4]) }
    Given(:b) { Set.new([3, 4, 5, 6]) }

    context '和集合' do
      let(:expected) { Set.new([1, 2, 3, 4, 5, 6]) }

      When(:result1) { a.union(b) }
      Then { result1 ==  expected }

      When(:result2) { a + b }
      Then { result2 == expected }

      When(:result3) { a | b }
      Then { result3 == expected }
    end

    context '積集合' do
      let(:expected) { Set.new([3, 4]) }

      When(:result1) { a.intersection(b) }
      Then { result1 == expected }

      When(:result2) { a & b }
      Then { result2 == expected }
    end

    context '差集合' do
      let(:expected) { Set.new([1, 2]) }

      When(:result1) { a.difference(b) }
      Then { result1 == expected }

      When(:result2) { a - b }
      Then { result2 == expected }
    end
  end
end
